File: Order.cs
using System.Runtime;
using System.Runtime.CompilerServices;
using Microsoft.VisualBasic;

class Order
{
    int orderNr { get; }
    Boolean isStudentOrder;
    private List<MovieTicket> _tickets = new List<MovieTicket>();

    public Order(int orderNr, Boolean isStudentOrder)
    {
        this.orderNr = orderNr;
        this.isStudentOrder = isStudentOrder;
    }

    public void addSeatReservation(MovieTicket ticket)
    {
        _tickets.Add(ticket);
    }

    public double calculatePrice()
    {
        double totalPrice = 0;
        int amountOfTickets = _tickets.Count;

        for (int i = 0; i < amountOfTickets; i++)
        {
            MovieTicket ticket = _tickets[i];
            MovieScreening screening = ticket.MovieScreening;

            bool isWeekday =
                screening.DateAndTime.DayOfWeek >= DayOfWeek.Monday &&
                screening.DateAndTime <= screening.DateAndTime.AddDays(0) &&
                screening.DateAndTime.DayOfWeek <= DayOfWeek.Thursday;

            bool secondTicketFree =
                i % 2 == 1 && (isStudentOrder || isWeekday);

            if (secondTicketFree)
            {
                continue;
            }

            double price = ticket.GetPrice(isStudentOrder);

            totalPrice += price;
        }

        bool isWeekend =
            _tickets[0].MovieScreening.DateAndTime.DayOfWeek == DayOfWeek.Friday ||
            _tickets[0].MovieScreening.DateAndTime.DayOfWeek == DayOfWeek.Saturday ||
            _tickets[0].MovieScreening.DateAndTime.DayOfWeek == DayOfWeek.Sunday;

        if (!isStudentOrder && isWeekend && amountOfTickets >= 6)
        {
            totalPrice *= 0.9;
        }

        return totalPrice;
    }


    public void export(TicketExportFormat format)
    {
        switch (format)
        {
            case TicketExportFormat.PLAINTEXT:
                Console.WriteLine($"Order #{orderNr}");
                Console.WriteLine(isStudentOrder ? "Student order" : "Regular order");
                Console.WriteLine("Tickets:");

                foreach (MovieTicket ticket in _tickets)
                {
                    Console.WriteLine(
                        $"- {ticket.MovieScreening.Movie} | " +
                        $"{ticket.MovieScreening.DateAndTime} | " +
                        $"Row {ticket.RowNr}, Seat {ticket.SeatNr}" +
                        (ticket.IsPremium ? " (Premium)" : "")
                    );
                }

                Console.WriteLine($"Total price: €{calculatePrice():0.00}");
                break;

            case TicketExportFormat.JSON:
                Console.WriteLine("{");
                Console.WriteLine($"  \"orderNr\": {orderNr},");
                Console.WriteLine($"  \"isStudentOrder\": {isStudentOrder.ToString().ToLower()},");
                Console.WriteLine($"  \"totalPrice\": {calculatePrice():0.00},");
                Console.WriteLine("  \"tickets\": [");

                for (int i = 0; i < _tickets.Count; i++)
                {
                    MovieTicket ticket = _tickets[i];

                    Console.WriteLine("    {");
                    Console.WriteLine($"      \"movie\": \"{ticket.MovieScreening.Movie}\",");
                    Console.WriteLine($"      \"dateTime\": \"{ticket.MovieScreening.DateAndTime}\",");
                    Console.WriteLine($"      \"row\": {ticket.RowNr},");
                    Console.WriteLine($"      \"seat\": {ticket.SeatNr},");
                    Console.WriteLine($"      \"premium\": {ticket.IsPremium.ToString().ToLower()}");
                    Console.Write("    }");

                    if (i < _tickets.Count - 1)
                        Console.WriteLine(",");
                    else
                        Console.WriteLine();
                }

                Console.WriteLine("  ]");
                Console.WriteLine("}");
                break;

            default:
                throw new ArgumentOutOfRangeException(nameof(format), format, null);
        }
    }

}

File: src/main/java/soagroup15a/Order.java
package soagroup15a;

import java.time.DayOfWeek;
import java.util.ArrayList;
import java.util.List;

public class Order {

    private int orderNr;
    private boolean isStudentOrder;
    private List<MovieTicket> tickets;

    public Order(int orderNr, boolean isStudentOrder) {
        this.orderNr = orderNr;
        this.isStudentOrder = isStudentOrder;
        this.tickets = new ArrayList<>();
    }

    public int getOrderNr() {
        return orderNr;
    }

    public void addSeatReservation(MovieTicket ticket) {
        tickets.add(ticket);
    }

    public double calculatePrice() {
        double totalPrice = 0.0;
        int ticketCount = tickets.size();

        for (int i = 0; i < ticketCount; i++) {
            MovieTicket ticket = tickets.get(i);
            MovieScreening screening = ticket.getMovieScreening();

            double price = screening.getPricePerSeat();

            if (ticket.isPremiumTicket()) {
                price += isStudentOrder ? 2.0 : 3.0;
            }

            DayOfWeek day = screening.dateAndTime.getDayOfWeek();
            boolean isWeekday = day == DayOfWeek.MONDAY || day == DayOfWeek.TUESDAY
                    || day == DayOfWeek.WEDNESDAY || day == DayOfWeek.THURSDAY;

            boolean secondTicketFree = (i % 2 == 1) && (isStudentOrder || isWeekday);

            if (!secondTicketFree) {
                totalPrice += price;
            }
        }

        if (!isStudentOrder && ticketCount >= 6) {
            boolean allWeekend = true;

            for (MovieTicket ticket : tickets) {
                DayOfWeek day = ticket.getMovieScreening().dateAndTime.getDayOfWeek();
                if (day != DayOfWeek.SATURDAY && day != DayOfWeek.SUNDAY) {
                    allWeekend = false;
                    break;
                }
            }

            if (allWeekend) {
                totalPrice *= 0.9;
            }
        }

        return totalPrice;
    }

    public String export(TicketExportFormat format) {
        StringBuilder sb = new StringBuilder();

        switch (format) {
            case PLAINTEXT:
                sb.append("Order Number: ").append(orderNr).append("\n");
                sb.append("Student Order: ")
                        .append(isStudentOrder ? "Yes" : "No").append("\n");
                sb.append("Tickets:\n");

                for (MovieTicket ticket : tickets) {
                    sb.append(ticket).append("\n");
                }

                sb.append(String.format("Total Price: €%.2f%n", calculatePrice()));
                break;

            case JSON:
                sb.append("{\n");
                sb.append("  \"orderNr\": ").append(orderNr).append(",\n");
                sb.append("  \"isStudentOrder\": ").append(isStudentOrder).append(",\n");

                sb.append("  \"tickets\": [\n");
                for (int i = 0; i < tickets.size(); i++) {
                    sb.append("    {\n");
                    sb.append("      \"description\": \"").append(tickets.get(i)).append("\"\n");
                    sb.append("    }");
                    if (i < tickets.size() - 1) {
                        sb.append(",");
                    }
                    sb.append("\n");
                }
                sb.append("  ],\n");

                sb.append(String.format("  \"totalPrice\": %.2f\n", calculatePrice()));
                sb.append("}");
                break;
        }

        return sb.toString();
    }
}
